ENH: Add Dask Array API support

lithomas1 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#26724

#### What does this implement/fix? Explain your changes.


#### Any other comments?

This depends on unmerged/unreleased changes in array-api-compat


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lithomas1 committed:
add all files

lithomas1 committed:
don't use np.asarray to force computation

lithomas1 committed:
some mods

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lithomas1 committed:
update

lithomas1 committed:
change back sign coercion stuff

lithomas1 committed:
update

lithomas1 committed:
avoid hang, but why?

lithomas1 committed:
fixes and remove linear model support

lithomas1 committed:
remove test notebook

lithomas1 commented on sklearn/discriminant_analysis.py:
```diff
@@ -534,7 +546,13 @@ def _solve_svd(self, X, y):
+
+        if isinstance(X, da.Array):
+            # dask can flip signs of Vt sometimes
+            _, S, Vt = svd(X, full_matrices=False, coerce_signs=False)
```
Not sure if the sign should matter in the result.

If it does, I can fix the svd inside array-api-compat to do ``coerce_signs=False``

lithomas1 commented on sklearn/discriminant_analysis.py:
```diff
@@ -113,6 +113,14 @@ def _class_means(X, y):
     xp, is_array_api_compliant = get_namespace(X)
     classes, y = xp.unique_inverse(y)
+    # Force lazy array api backends to call compute
+    if hasattr(classes, "persist"):
```
I'm planning on extracting the calls here to a ``compute`` function to realize lazy arrays.

Not sure if it will live in scikit-learn, though. I think I'd want it to live in array-api-compat, but I haven't discussed this with the folks there yet.

Based on usage here, the new ``compute`` function could have an option to e.g. compute shape only if that's what's needed, or compute the full array.

betatim replied:
I think this kind of code needs to go somewhere else. Having the "lazyness" leak through is a bit silly. I'm not sure where it should be though. 

I don't know how to resolve this issue. In the past I've suggested that accessing something like `.shape` should just trigger the computation for you or that everything needs to grow a `.compute`, even eager libs. Both would allow array consumers like scikit-learn to write code that does not care whether the implementation is lazy or not. However there doesn't seem to be much support for this within the Array API community. The alternative is having to place code that checks "is this lazy? if yes trigger compute" in consumers like scikit-learn which I think is not great.

So this is maybe not a task for you alone to solve @lithomas1 but we need to find some kind of solution

lithomas1 replied:
Thinking about this some more, would scikit-learn be happy with a ``shape`` helper (as opposed to my earlier suggestion of a ``compute`` helper) from array-api-compat?

This would handle the laziness of the various array libraries out there and call ``compute`` (or whatever the equivalent is) on the array if the shape needs to be materialized.

(In the event that array-api-compat is not installed, we can just define this to return, e.g. ``x.shape`` like scikit-learn already does)

The advantage of this approach would be that scikit-learn doesn't have to think about laziness anymore - that work would be outsourced to array-api-compat. 

The only downsides of this approach of this approach are that
- We indiscriminately materialize (for lazy arrays) even if it's strictly not necessary (e.g. we just access the shape to pass it to an array constructor like np.zeros). I don't think we'll lose too much (if any) performance here, though.
- scikit-learn devs need to remember that accessing ``.shape`` is banned, and existing usages have to be migrated.
  - This can be mitigated with a pre-commit hook, to automatically detect ``.shape`` accesses.
I think with something like ruff, one can write a custom rule to automatically rewrite ``.shape`` accesses to ``shape(x)``
  - For existing usages, this is something that can be done as part of the Array API migration process, so it shouldn't cause too much churn on its own

How does this sound to you?

lithomas1 replied:
@betatim any thoughts on the above?

ogrisel replied:
Note that a `compute` helper could also help deal with the boolean-masked assignment problem in `r2_score` described in more details in this comment:

https://github.com/scikit-learn/scikit-learn/pull/28588#issuecomment-2031595468

This also a lazy evaluation problem but not related to shape values.

ogrisel replied:
> I think having a helper like shape() is the only(?) way forward for now. I'd not add it to array-api-compat but instead add it to scikit-learn in utils/_array_api.py - we already have a few helpers there for things that feel "scikit-learn specific".

We could also do both: have a public helper in `array-api-compat` and a private scikit-learn specific helper in scikit-learn, that does nothing for libraries that are not accessed via array-api-compat (e.g. `array-api-strict`) as long the spec does not provide a standard way to deal with this.

ogrisel replied:
> More adventurous: I wonder if we can even wrap the dask namespace (and via that its arrays) to make it so what .shape access triggers the computation. That way people who edit scikit-learn's code base don't need to know anything about this issue.

Not sure how feasible this is and whether its desirable or not to trigger computation implicitly when using lazy libraries.

lithomas1 commented on sklearn/discriminant_analysis.py:
```diff
@@ -113,6 +113,14 @@ def _class_means(X, y):
+        # compute right now
+        # Probably a dask bug
+        # (the error is also kinda flaky)
+        y = y.compute()
```
To shed more light on what this is, I think this bug happens when scikit-learn calls unique_inverse.

I think something is going wrong in dask somewhere where the results of intermediate operations are getting corrupted.

When the error occurs in computation is somewhat flaky, but it happens more often than not without the compute here.

ogrisel replied:
Did you investigate the cause of the corruption? It might be worth reporting a minimal reproducer upstream.

lithomas1 commented on sklearn/utils/_array_api.py:
```diff
@@ -87,7 +88,7 @@ def device(x):
         `device` object (see the "Device Support" section of the array API spec).
     """
-    if isinstance(x, (numpy.ndarray, numpy.generic)):
+    if isinstance(x, (numpy.ndarray, numpy.generic)) or not hasattr(x, "device"):
```
Might be able to use the ``device`` helper from ``array-api-compat`` instead.

lithomas1 commented:
While LDA was a bit hard to port over to dask, PCA worked perfectly out of the box!
(which I think is a pretty big win for the array API)

The other preprocessing/metrics/tools also worked out of the box, I think (at least judging by the tests).

lithomas1 commented:
R.e. performance

**Testing out of core**

for dask generated by dask-ml with  parameters
```
n_samples=100_000
n_classes=2
n_informative=5
```

on a Gitpod machine with 2 cores and 8 GB RAM, I get

14m51s for 100,000,000 by 20 LDA (14.90 GB) # 100,000,000 samples
47 seconds for 10,000,000 by 20 LDA (1.49 GB) # 10,000,000 samples
6m 3.5s for 50,000,000 by 20 LDA (7.45 GB).    # 50,000,000 samples

which I think is a pretty decent scaling

** Distributed computation **

Still benchmarking.

adrinjalali commented:
cc @betatim @ogrisel

betatim commented on sklearn/discriminant_analysis.py:
```diff
@@ -534,7 +546,13 @@ def _solve_svd(self, X, y):
         # Use SVD to find projection in the space spanned by the
         # (n_classes) centers
-        _, S, Vt = svd(X, full_matrices=False)
+        import dask.array as da
```
We don't want to explicitly import array libraries or have code that checks the type in general code like this. If we need to we can add helpers in `utils/_array_api.py` to deal with differences that remain after using the Array API. Though even there we can't assume that `import dask` will work, as it isn't a dependency of scikit-learn, so we need a way to determine that dask is installed before importing it or accessing it in a different way

lithomas1 replied:
I left a comment about this above https://github.com/scikit-learn/scikit-learn/pull/28588#discussion_r1515308238.

Basically, the signs on the elements in the results are sometimes flipped with dask's svd, and I wasn't sure if users of scikit-learn expect this. 

The closest thing I can find was this
https://stackoverflow.com/questions/44765682/in-sklearn-decomposition-pca-why-are-components-negative

In any case, I will clean up this PR so that either (1) this happens implicitly in array-api-compat or (2) if the code block is unnecessary, I'll just delete it.

(so we won't end up with any explicit imports in the final code).

lithomas1 replied:
I removed these changs.

lithomas1 committed:
skip some tests for PCA

lithomas1 commented:
FYI, array-api-compat fixes are ongoing here https://github.com/data-apis/array-api-compat/pull/110

lithomas1 committed:
update

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lithomas1 committed:
add min version check for array-api-compat

lithomas1 committed:
patches

lithomas1 committed:
bump array-api-compat

lithomas1 committed:
fix r2_score

lithomas1 committed:
fix tests

lithomas1 commented:
Since array-api-compat 1.5.1 came out and CI is green here,

I'm going to be marking this PR as ready for review.

The only other change I'm planning right now is splitting out the LDA changes, since that requires a patch to dask itself.

The correct way to handle laziness is also something that might be good to think about. 
(It might be good to loop in more scikit-learn devs about this).

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lithomas1 committed:
remove dask specific changes

lithomas1 committed:
Merge branch 'main' into wip-dask-array-api

ogrisel commented on sklearn/decomposition/tests/test_pca.py:
```diff
@@ -891,8 +891,16 @@ def check_array_api_get_precision(name, estimator, array_namespace, device, dtyp
 ):
+    # Dask doesn't implement slogdet from the Array API
+    # which is used in score/score_samples
+    skip_methods = {"dask.array": ["score", "score_samples"]}
```
Could you please add a TODO comment with a link to an upstream issue, either in `dask` or in `array-api-compat` to track the resolution of this problem?

ogrisel replied:
Actually, wouldn't it make sense to import `_array_api_skip_methods` from `sklearn.utils._array_api` instead of defining this dict inline here?

If so, lets move the comment there.

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -1303,8 +1303,10 @@ def _check_y(y, multi_output=False, y_numeric=False, estimator=None):
+    if y_numeric:
+        xp, _ = get_namespace(y)
+        if not xp.isdtype(y.dtype, "numeric"):
+            y = xp.astype(y, xp.float64)
```
Out of curiosity, which test(s) revealed this missing code update? It's weird that we did not catch it earlier with `array-api-strict` or PyTorch.

lithomas1 replied:
I think this crept in with some other changes to the estimators that I was making (unrelated to dask stuff), sorry.

I've taken this out (but yes, this should be needed int he future when some other estimators are ported, I forgot which one).

ogrisel commented:
Thanks @lithomas1! Could you please add an entry referencing this PR in `doc/whats_new/v1.5.rst`? There is a dedicated section for Array API changes. This should fix the "Check Changelog" CI failure.

ogrisel commented:
> The only other change I'm planning right now is splitting out the LDA changes, since that requires a patch to dask itself.

Sounds good, however in the last CI run with dask I see 3 failures:

```
FAILED ::__init__.py::__init__.py::test_common.py::test_array_api_compliance[r2_score-check_array_api_regression_metric-dask.array-None-None] - ValueError: Boolean index assignment in Dask expects equally shaped arrays.
FAILED ::__init__.py::test_common.py::test_estimators[LinearDiscriminantAnalysis()-check_array_api_input(array_namespace=dask.array,dtype_name=None,device=None,skip_methods={})] - ValueError: An error occurred while calling the wrap_func_shape_as_first_ar...
FAILED ::__init__.py::__init__.py::test_multiclass.py::test_is_multilabel_array_api_compliance[dask.array-None-None] - AssertionError: is_multilabel(dask.array<array, shape=(10, 10), dtype=int64...
```

only one of them  seems directly related to the lazy shape needed for `LinearDiscriminantAnalysis`.

The failures for `check_array_api_regression_metric` with `r2_score` and `is_multilabel` seem to be from a different nature but I would need to investigate with a debugger to confirm.

> The correct way to handle laziness is also something that might be good to think about.
(It might be good to loop in more scikit-learn devs about this).

I am personally fine with some kind of "trigger compute now if lazy" help in array-api-compat as a short term solution to be able to experiment with lazy Array API libs in scikit-learn and see empirically how much we have to rely on it in scikit-learn to inform the decision of a future version of the SPEC itself to better explicitly handle lazyness with an official API.

ogrisel reviewed (commented):
Some inline feedback:

ogrisel commented:
Here is the info about the r2 score metric. So it's indeed a with of lazy shape:

```
> /Users/ogrisel/code/scikit-learn/sklearn/metrics/_regression.py(884)_assemble_r2_explained_variance()
-> output_scores[valid_score] = 1 - (
(Pdb) l
879             # (note: even if denominator is zero, thus avoiding NaN scores)
880             output_scores = xp.ones([n_outputs], device=device, dtype=dtype)
881             # Non-zero Numerator and Non-zero Denominator: use the formula
882             valid_score = nonzero_denominator & nonzero_numerator
883  
884  ->         output_scores[valid_score] = 1 - (
885                 numerator[valid_score] / denominator[valid_score]
886             )
887  
888             # Non-zero Numerator and Zero Denominator:
889             # arbitrary set to 0.0 to avoid -inf scores
(Pdb) p valid_score
dask.array<and_, shape=(2,), dtype=bool, chunksize=(2,), chunktype=numpy.ndarray>
(Pdb) p numerator
dask.array<sum-aggregate, shape=(2,), dtype=float64, chunksize=(2,), chunktype=numpy.ndarray>
(Pdb) p denominator
dask.array<sum-aggregate, shape=(2,), dtype=float64, chunksize=(2,), chunktype=numpy.ndarray>
(Pdb) p numerator[valid_score]
dask.array<getitem, shape=(nan,), dtype=float64, chunksize=(nan,), chunktype=numpy.ndarray>
(Pdb) p 1 - (numerator[valid_score] / denominator[valid_score])
dask.array<sub, shape=(nan,), dtype=float64, chunksize=(nan,), chunktype=numpy.ndarray>
(Pdb) p output_scores
dask.array<ones_like, shape=(2,), dtype=float64, chunksize=(2,), chunktype=numpy.ndarray>
```

This causes:

```
E               ValueError: Boolean index assignment in Dask expects equally shaped arrays.
E               Example: da1[da2] = da3 where da1.shape == (4,), da2.shape == (4,) and da3.shape == (4,).
E               Alternatively, you can use the extended API that supportsindexing with tuples.
E               Example: da1[(da2,)] = da3.
```

I wonder what should be done in this case. Trigger an explicit computation prior to assignment?

ogrisel commented:
For information, the tuple indexing syntax suggested in the error message would not work for this particular case:

```diff
diff --git a/sklearn/metrics/_regression.py b/sklearn/metrics/_regression.py
index ad5c76810f..f0931045ab 100644
--- a/sklearn/metrics/_regression.py
+++ b/sklearn/metrics/_regression.py
@@ -881,7 +881,7 @@ def _assemble_r2_explained_variance(
         # Non-zero Numerator and Non-zero Denominator: use the formula
         valid_score = nonzero_denominator & nonzero_numerator
 
-        output_scores[valid_score] = 1 - (
+        output_scores[(valid_score,)] = 1 - (
             numerator[valid_score] / denominator[valid_score]
         )
```

would lead to:

```
E               ValueError: Arrays chunk sizes are unknown: (nan,)
E               
E               A possible solution: https://docs.dask.org/en/latest/array-chunks.html#unknown-chunks
E               Summary: to compute chunks sizes, use
E               
E                  x.compute_chunk_sizes()  # for Dask Array `x`
E                  ddf.to_dask_array(lengths=True)  # for Dask DataFrame `ddf`
```

ogrisel commented:
So to conclude we have identified only two types of problems related to the lazy evaluation in dask:

- decisions in the code based on lazy shape item values;
- masked assignments with the same lazy boolean array on each side.

Shall we create two issues upstream and link to this PR?

In the mean time we could either:

- introduce a *temporary*, dask-specific eager evaluation private helper in scikit-learn and use it sparingly whenever we need value-dependent branching in our code;
- introduce such a public helper in `array-api-compat` and use it in scikit-learn to unblock this PR.

lithomas1 commented:
> > The only other change I'm planning right now is splitting out the LDA changes, since that requires a patch to dask itself.
> 
> Sounds good, however in the last CI run with dask I see 3 failures:
> 
> ```
> FAILED ::__init__.py::__init__.py::test_common.py::test_array_api_compliance[r2_score-check_array_api_regression_metric-dask.array-None-None] - ValueError: Boolean index assignment in Dask expects equally shaped arrays.
> FAILED ::__init__.py::test_common.py::test_estimators[LinearDiscriminantAnalysis()-check_array_api_input(array_namespace=dask.array,dtype_name=None,device=None,skip_methods={})] - ValueError: An error occurred while calling the wrap_func_shape_as_first_ar...
> FAILED ::__init__.py::__init__.py::test_multiclass.py::test_is_multilabel_array_api_compliance[dask.array-None-None] - AssertionError: is_multilabel(dask.array<array, shape=(10, 10), dtype=int64...
> ```
> 
> only one of them seems directly related to the lazy shape needed for `LinearDiscriminantAnalysis`.

Sorry, I pulled out all my changes last night related to the triggering compute on lazy shapes intentionally (and went to bed before leaving a comment ...), since I realized both of them depend on changing dask behavior/fixing bugs in dask.

For LDA, there's an annoying race condition, that if I was able to patch in dask, we would be able to use the ``shape`` helper, I mentioned above.

``r2_score``/``is_multilabel`` has the issue where when dask is doing boolean assignment with arrays with a lazy shape, it doesn't automatically call compute on the operands, and errors out with a shape doesn't match error. 
So, we would have to insert a ``shape`` helper call randomly, which would go against the idea of being able to write code for both lazy and eager array implementations, and not having to worry about the laziness.

In this case, we might have to go for the heavier approach that @betatim mentioned of wrapping the dask array class itself, and call compute on lazy arrays in ``__setitem__`` (and other methods if necessary) before handing the ``__setitem__`` of to dask.

(I haven't worked out the details of this yet though, but my first thought is that we can return wrapped arrays from the ``array_namespace`` call)


IMO, right now, having these two things not work is not really a big deal for dask support in scikit-learn (esp. since the estimators/methods in progress right now, e.g. #27800 for ``RidgeCV``, and #28106 for ``mean_tweedie_deviance`` look like they should work out of the box without any changes at an initial glance)

So, 
@betatim @ogrisel 
What do you think about just adding skips if necessary to any methods/estimators that dask doesn't support out of the box, and collecting data in a doc/Github issue somewhere, so we can have a more complete picture of what's needed to support laziness?

(I'm happy to transcribe our conversations on this topic (and planned solutions) so that we don't lose it.)

ogrisel commented:
I am fine with this if you find a concise way to add the necessary skips with links to one or more dedicated issues in either scikit-learn or upstream issue trackers.

lithomas1 committed:
modify test machinery

lithomas1 committed:
add whatsnew note

lithomas1 committed:
more test changes

lithomas1 committed:
revert more changes

lithomas1 committed:
fix last test

lithomas1 commented:
This should be ready for another pass now.

Primary mechanism for skips would be the ``_array_api_skips`` dictionary in ``sklearn/utils/_array_api.py``

I also added a ``skip_dask`` fixture/decorator for individual tests that test for ``array_api_compliance``.

ogrisel commented on sklearn/conftest.py:
```diff
@@ -313,3 +313,19 @@ def print_changed_only_false():
+    """
+    Xfails an array API compliance test for dask.
+    (i.e. when the array_namespace fixture yields 'dask.array')
+    """
```
```suggestion

    When using this fixture, please insert a comment to explain what particular
    aspect of Dask is preventing the tests to pass, e.g. missing module level
    function or array level method, value-dependent shape item values or array
    assignments with a value-dependent boolean mask.
    """
```

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -10,6 +10,20 @@
+    # Dask doesn't implement slogdet from the Array API
+    # which is used in score/score_samples
+    "PCA": {"dask.array": ["score", "score_samples"]},
+    "LinearDiscriminantAnalysis": {"dask.array": "all"},
```
```suggestion
    # Lazy evaluation semantics: value-dependent shape item value (nan):
    "LinearDiscriminantAnalysis": {"dask.array": "all"},
```

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -10,6 +10,20 @@
+_array_api_skips = {
+    # Dask doesn't implement slogdet from the Array API
+    # which is used in score/score_samples
+    "PCA": {"dask.array": ["score", "score_samples"]},
+    "LinearDiscriminantAnalysis": {"dask.array": "all"},
+    "r2_score": {"dask.array": "all"},
```
```suggestion
    "PCA": {"dask.array": ["score", "score_samples"]},
    "LinearDiscriminantAnalysis": {"dask.array": "all"},
    # Lazy evaluation semantics: cannot assign to an array using a value-dependent
    # boolean mask.
    "r2_score": {"dask.array": "all"},
```

lithomas1 replied:
applied.

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -947,10 +957,15 @@ def check_array_api_input(
-            continue
+        if method is None or method_name in methods_to_skip:
+            raise SkipTest(
+                f"{array_namespace} is not array-API compliant for method"
```
I think we tend to use the title case convention when mentioning the Array API standard.

```suggestion
                f"{array_namespace} is not Array API compliant for method"
```

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -886,7 +889,14 @@ def check_array_api_input(
+    (to skip all methods for this estimator)
     """
+    if skips.get(array_namespace) == "all":
+        raise SkipTest(f"{array_namespace} is not array-API compliant for {name}")
```
```suggestion
        raise SkipTest(f"{array_namespace} is not Array API compliant for {name}")
```

ogrisel commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1736,6 +1737,11 @@ def test_metrics_pos_label_error_str(metric, y_pred_threshold, dtype_y_str):
 ):
+    if _array_api_skips.get(metric.__name__, {}).get(array_namespace) == "all":
+        pytest.skip(
+            f"{array_namespace} is not array-API compliant for {metric.__name__}"
```
```suggestion
            f"{array_namespace} is not Array API compliant for {metric.__name__}"
```

ogrisel commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -378,6 +380,7 @@ def test_is_multilabel():
             ), f"is_multilabel({example!r}) should be {dense_exp}"
 
 
+@skip_dask
```
```suggestion
@skip_dask  # lazy evaluation semantics for value-dependent shape values 
```

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -74,6 +89,13 @@ def _check_array_api_dispatch(array_api_dispatch):
                 " specification"
             )
+        array_api_compat_version = parse_version(array_api_compat.__version__)
+        min_array_api_compat_version = "1.5.1"
```
Let's also add this requirement to `sklearn/_min_dependencies.py` using the "tests" scope (as we do for other soft dependencies such as pandas, polars or matplotlib).

Then this line can be updated by looking up the version number from `sklearn._min_dependencies.dependent_packages`.

ogrisel reviewed (commented):
Here is another pass of feedback, possibly final on my end.

Would it be possible to open issues upstream, either in the dask, array-api-compat repos or the the Array API spec repo to link to from the comments of the skipped tests?

ogrisel commented:
BTW, I launched the tests on hosts with cuda and MPS devices to confirm that this PR did not introduce any regression for those.

lithomas1 committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lithomas1 committed:
rest of suggestions

lithomas1 committed:
address rest of comments

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lithomas1 committed:
add array-api-compat to pyproject.toml

lithomas1 committed:
fix linter

lithomas1 commented:
> Here is another pass of feedback, possibly final on my end.
> 
> Would it be possible to open issues upstream, either in the dask, array-api-compat repos or the the Array API spec repo to link to from the comments of the skipped tests?

Yep, I'm tracking failures in https://github.com/scikit-learn/scikit-learn/issues/26724#issuecomment-2038674794

Failures are listed as checkboxes (with accompanying upstream issues if necessary).

Thanks for the reviews.
I've updated the PR following the feedback.

ogrisel commented on doc/modules/array_api.rst:
```diff
@@ -25,9 +25,12 @@ At this stage, this support is **considered experimental** and must be enabled
 
+    `dask.array` support is incomplete at the time of writing: some methods and
+    estimators many not work while we work out a way to handle Dask's lazy evaluation
+    semantics in a library-agnostic way.
```
The missing blank line in my suggested change probably broke the doc CI with:

```
/home/circleci/project/doc/modules/array_api.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

```suggestion
    semantics in a library-agnostic way.

```

ogrisel commented on doc/modules/array_api.rst:
```diff
@@ -25,9 +25,12 @@ At this stage, this support is **considered experimental** and must be enabled
     are known to work with scikit-learn's estimators.
 
+    `dask.array` support is incomplete at the time of writing: some methods and
+    estimators many not work while we work out a way to handle Dask's lazy evaluation
```
```suggestion
    estimators may not work while we work out a way to handle Dask's lazy evaluation
```

ogrisel committed:
Fix doc

ogrisel reviewed (approved):
Assuming the CI is green, LGTM for a first iteration.

What are you thoughts @betatim? Are you comfortable with merging partial dask support with skipped tests? Or do you prefer to implement a more complete support first?

betatim commented on doc/whats_new/v1.5.rst:
```diff
@@ -47,6 +47,12 @@ See :ref:`array_api` for more details.
 
+**Libraries:**
+
+- ``dask.array`` is now experimentally supported as an array API backend.
```
```suggestion
- ``dask.array`` is now experimentally supported as an Array API backend.
```

betatim replied:
Looks like Olivier commented on a few of these before as well. I won't comment all that I find, but it would be great to have "Array API" every where.

lucascolley replied:
not to be pedantic, but I would recommend _not_ capitalising this as per https://github.com/data-apis/array-api/discussions/778 !

ogrisel replied:
Thanks for the feedback, I was not aware of this. I guess we have a lot of doc/comments to update...

Let's fix that in a dedicated PR.

betatim commented:
> Are you comfortable with merging partial dask support with skipped tests? Or do you prefer to implement a more complete support first?

Unsure. I started looked at the diff just now. I think if some estimators work and some don't that is one thing. But if for an estimator only some of it works, that is a bit weird. From a user's perspective it would be weird I think because you can't really do much if only part of an estimator works.

betatim commented on sklearn/conftest.py:
```diff
@@ -313,3 +313,24 @@ def print_changed_only_false():
+    if array_namespace == "dask.array":
+        pytest.skip(
+            reason=(
+                "Estimator/method does not work because of dask array API compliance"
+                " issues"
```
```suggestion
                "Estimator/method does not work because of missing dask Array API compliance."
```
WDYT?

betatim commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1736,6 +1737,11 @@ def test_metrics_pos_label_error_str(metric, y_pred_threshold, dtype_y_str):
 ):
+    if _array_api_skips.get(metric.__name__, {}).get(array_namespace) == "all":
+        pytest.skip(
+            f"{array_namespace} is not Array API compliant for {metric.__name__}"
```
```suggestion
            f"Skipping {metric.__name__} because of missing Array API compliance in {array_namespace}"
```
It feels like the current text is backwards. What do you think of this change? Trying to make it clear that "we are skipping testing for X because something in {array_namespace}'s Array API support is missing" - which is what I think we are doing.

betatim commented on sklearn/utils/_array_api.py:
```diff
@@ -74,6 +93,13 @@ def _check_array_api_dispatch(array_api_dispatch):
                 "array_api_compat is required to dispatch arrays using the API"
```
```suggestion
                "array-api-compat is required to dispatch arrays using the API"
```
Trying to be consistent with the spelling below. The package on PyPI is called "array-api-compat" so I think it makes sense to use that

betatim commented on sklearn/utils/estimator_checks.py:
```diff
@@ -946,10 +956,15 @@ def check_array_api_input(
         "transform",
     )
 
+    methods_to_skip = skips.get(array_namespace, [])
+
     for method_name in methods:
         method = getattr(est, method_name, None)
-        if method is None:
-            continue
+        if method is None or method_name in methods_to_skip:
+            raise SkipTest(
+                f"{array_namespace} is not Array API compliant for method"
+                f" {method_name} of {name}"
+            )
```
Why the change for `method is None`? This is the case when an estimator doesn't have a particular method, in which case we aren't skipping a test (and certainly not because of missing Array API support :D)

betatim commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1019,6 +1034,7 @@ def check_array_api_input_and_values(
     array_namespace,
     device=None,
     dtype_name="float64",
+    skip_methods={},
```
I think we should use `skip_methods=None` + `if skip_methods is None: skip_methods = {}` in the function body. Mutable values as default value for a function argument is a trap waiting for someone to fall into.

<details>
<summary>Details</summary>

```python
def foo(x, a=[]):
  a.append(x)
  print(a)

foo(1)  # prints [1]
foo(1)  # print [1, 1]
```

</details>

betatim commented:
Can we use array-api-compat to add `slogdet` to dask arrays? Then we'd avoid the problem that an estimator semi works.

For the two other skips (`LinearDiscriminantAnalysis` and `r2_score`), what is the error/exception a user will see when they try to use them with a dask array? Should we have a dedicated exception message for this case or is "random error", like you get when you use an estimator that doesn't yet have Array API support, good enough?

ogrisel commented:
> Should we have a dedicated exception message for this case or is "random error", like you get when you use an estimator that doesn't yet have Array API support, good enough?

This is going to be hard (impossible) to do that in a library agnostic way. Note that we already get low-level error message with dask array in `main`:


```python
>>> import sklearn
>>> sklearn.set_config(array_api_dispatch=True)
>>> import dask.array as da
>>> X = da.random.normal(size=(1000, 10))
>>> y = da.random.randint(0, 2, size=1000)
>>> from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
>>> LinearDiscriminantAnalysis().fit(X, y)
Traceback (most recent call last):
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/dask/backends.py:140 in wrapper
    return func(*args, **kwargs)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/dask/array/wrap.py:63 in wrap_func_shape_as_first_arg
    parsed = _parse_wrap_args(func, args, kwargs, shape)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/dask/array/wrap.py:33 in _parse_wrap_args
    chunks = normalize_chunks(chunks, shape, dtype=dtype)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/dask/array/core.py:3095 in normalize_chunks
    chunks = auto_chunks(chunks, shape, limit, dtype, previous_chunks)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/dask/array/core.py:3210 in auto_chunks
    raise ValueError(
ValueError: Can not perform automatic rechunking with unknown (nan) chunk sizes.

A possible solution: https://docs.dask.org/en/latest/array-chunks.html#unknown-chunks
Summary: to compute chunks sizes, use

   x.compute_chunk_sizes()  # for Dask Array `x`
   ddf.to_dask_array(lengths=True)  # for Dask DataFrame `ddf`

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  Cell In[11], line 1
    LinearDiscriminantAnalysis().fit(X, y)
  File ~/code/scikit-learn/sklearn/base.py:1474 in wrapper
    return fit_method(estimator, *args, **kwargs)
  File ~/code/scikit-learn/sklearn/discriminant_analysis.py:638 in fit
    self._solve_svd(X, y)
  File ~/code/scikit-learn/sklearn/discriminant_analysis.py:510 in _solve_svd
    self.means_ = _class_means(X, y)
  File ~/code/scikit-learn/sklearn/discriminant_analysis.py:116 in _class_means
    means = xp.zeros((classes.shape[0], X.shape[1]), device=device(X), dtype=X.dtype)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/array_api_compat/_internal.py:28 in wrapped_f
    return f(*args, xp=xp, **kwargs)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/array_api_compat/common/_aliases.py:134 in zeros
    return xp.zeros(shape, dtype=dtype, **kwargs)
  File ~/miniforge3/envs/dev/lib/python3.11/site-packages/dask/backends.py:142 in wrapper
    raise type(e)(
ValueError: An error occurred while calling the wrap_func_shape_as_first_arg method registered to the numpy backend.
Original Message: Can not perform automatic rechunking with unknown (nan) chunk sizes.

A possible solution: https://docs.dask.org/en/latest/array-chunks.html#unknown-chunks
Summary: to compute chunks sizes, use

   x.compute_chunk_sizes()  # for Dask Array `x`
   ddf.to_dask_array(lengths=True)  # for Dask DataFrame `ddf`
```

So from this point of view, this PR is not making things worse than they currently are. At least we document that dask support is not fully functional yet.

lithomas1 commented:
> Can we use array-api-compat to add `slogdet` to dask arrays? Then we'd avoid the problem that an estimator semi works.

This is not possible yet.

Dask doesn't expose a way to do a determinant yet. 
(and there doesn't seem to be an easy workaround, since dask's LU/QR decomps don't provide enough information to do it. I've opened an issue on the dask side, but there doesn't seem to be a response yet https://github.com/dask/dask/issues/11042)

I agree that it's not ideal to have dask not be able to use ``score``, but this is something that I think is reasonable to have users work around for now, e.g. by using ``_estimator_with_converted_arrays``, to convert the estimator from dask to numpy arrays.
(similar to how we transfer arrays from GPU to CPU on cupy)

I'm not sure how important ``score`` is though. There seem to be some usages of it, but curiously there seem to be no examples in scikit-learn itself on using score.

https://github.com/search?q=repo%3Ascikit-learn%2Fscikit-learn+pca.score&type=code

betatim commented:
My feeling overall is that adding dask support leaves us with something that is only half finished for the (very few) existing estimators. My first reaction to that is to wait a bit until Array API support in dask has progressed a bit more. My assumption is that there is no need to be in a hurry here.

ogrisel commented:
> I'm not sure how important score is though. There seem to be some usages of it, but curiously there seem to be no examples in scikit-learn itself on using score.

The score method is implicitly used by tools such as `cross_val_score` or `GridSearchCV`. However, it is true that it is very rarely used for `PCA` alone in practice. It's mostly used for supervised learning pipelines.

> I agree that it's not ideal to have dask not be able to use score, but this is something that I think is reasonable to have users work around for now, e.g. by using _estimator_with_converted_arrays, to convert the estimator from dask to numpy arrays.
(similar to how we transfer arrays from GPU to CPU on cupy)

It might also be a case the score method itself of the estimator can convert arrays to numpy if the namespace does not provide the necessary `xp.linalg.slogdet` method. For truncated PCA, we can expect the call of the `fit` method and maybe `get_precision` to be slow and would deserve running on accelerated namespace while the final `xp.linalg.slogdet` call should not be a performance critical operation (and the result is a scalar).

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lithomas1 committed:
fix ridgeregression

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

samir-nasibli commented on doc/whats_new/v1.5.rst:
```diff
@@ -87,6 +87,12 @@ See :ref:`array_api` for more details.
 
 **Classes:**
 
+<<<<<<< HEAD
```
```suggestion
```

ogrisel committed:
Remove conflict resolution marker

Co-authored-by: Samir Nasibli <samir.nasibli@intel.com>

ogrisel commented on sklearn/linear_model/_ridge.py:
```diff
@@ -288,11 +288,14 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None, copy=False):
     U, s, Vt = xp.linalg.svd(X, full_matrices=False)
-    idx = s > 1e-15  # same default value as scipy.linalg.pinv
-    s_nnz = s[idx][:, None]
+    idx = s > 1e-15[:, None]  # same default value as scipy.linalg.pinv
```
`>` does not have precedence of over `[]`, right? Trying this pattern locally yields:

```python
TypeError: 'float' object is not subscriptable
```

I think you meant the following instead:

```suggestion
    # scipy.linalg.pinv also thresholds at 1e-15 by default.
    idx = (s > 1e-15)[:, None]
```

ogrisel replied:
But then the following `s = s[:, None]` seems redundant and similarly for `idx[:, None]` in the call to `where`.

Also we should probably rename `idx` to something more correct and explicit such as `strictly_positive_mask`.

lithomas1 replied:
Thanks! I think I had this working locally and passing tests, but I must have messed something up on the merge.

I'm planning on circling back to this towards the end of summer, so feel free to take over this if you're interested in the meantime.

Not holding my breath, but also hoping that dask.array support improves in the meantime as well. 

(The recent Array API updates suggest that sort is the most pressing thing that's missing in dask. Linalg wise, the eig family of methods is probably the next biggest missing feature in dask, we just haven't seen it come up yet since not a lot of estimators have been ported yet).

ogrisel commented:
Let's convert to draft for the time being then.

ogrisel committed:
Merge main and regenerate lock files

ogrisel reviewed (changes requested):
I merged `main` to retrigger a round of CI with the current state of scikit-learn array API so that we have a better understanding on how much value dependent shapes / assignments are a problem and maybe compare the test results to the failures observed when attempting to run the same tests with jax: #29647.

I am changing my review state to request changes to avoid implying that this draft PR is ready to merge as is. We first need to agree if we accept partial dask support of if want to iron out all the problems induced by the lazy evaluation semantics before considering a merge of this PR.

lithomas1 commented:
Thanks for updating.

I'm back to working on this again now that my summer is over.

Last time I remember, I think had ridge regression working locally (after wrangling some issues with data-dependent output shapes which I think I worked around with where).

I'm going to try to fix some more issues in dask upstream like the lack of sorting, which seems to be a majority of failures, an also linalg.

r.e. dynamic shapes:
Have you tried running sklearn functions through jax with JIT on?
I believe dynamic shapes isn't an issue in jax without the JIT.

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

ogrisel commented:
> Have you tried running sklearn functions through jax with JIT on?

No I have not tried.

ogrisel commented:
For the record, #30340 was recently merged to `main` so this might unblock this PR because we can now leverage dask compat features implemented in `array-api-extra`.

lithomas1 commented:
Thanks, I'll rebase this this week and let you know how it goes.

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lithomas1 committed:
more fixes

lithomas1 committed:
Merge branch 'main' into wip-dask-array-api

lithomas1 commented:
Made a bit of progress.
Some notes I made along the way are:

TODOs
- [ ] Upstream fill_diagonal to array-api-extra (5 uses in sklearn)
- [ ] Upstream isin/in1d to array-api-extra (used in encoders)

Changes
- had to disable shape check in `_average` for lazy backends

Blockers
- [ ] setdiff1d in array-api-extra only works on eager arrays
	- Probably the next major thing to work on (this blocks LabelEncoder from working)
		- remaining 13 metrics tests also depend on this

Progress
- Fixed PCA (was broken due to fill_diagonal)
- Fixed the input validators (e.g. `is_multilabel`)
	- nunique in array-api-extra is very helpful so far
		- Able to do `int(nunique(...` to get n_classes
		- This forces compute (but shouldn't be so bad if we're only doing this on y)
	- Going to try to apply this to LDA later on
Thoughts
- nunique works, but it'd be better if it was part of a unique function though
	- Right now, we have to do double the unique computation, once on finding the unique values and once for nunique
- If we want to support lazy backends, there needs to be some thought as to how error checking should work
	- Right now, I disable them since for dask we don't know the shape until we compute. The alternative would be to defer them to raise during compute time.

lithomas1 committed:
WIP for LDA working

lithomas1 committed:
Merge branch 'wip-dask-array-api' of github.com:lithomas1/scikit-learn into wip-dask-array-api

lithomas1 committed:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into wip-dask-array-api

lucascolley commented:
> setdiff1d in array-api-extra only works on eager arrays

there was a PR in progress for this at https://github.com/data-apis/array-api-extra/pull/124, feel free to comment over there if that is blocking you

crusaderky commented:
> > setdiff1d in array-api-extra only works on eager arrays

To be more precise,
`setdiff1d(assume_unique=False)` only works for eager arrays.
`setdiff1d(assume_unique=True)` works fine in Dask (but not jax.jit) as long as the inputs' shapes do not contain NaNs.

ogrisel commented:
For the record, dask support in SciPy via array API is being deprioritized (at least in the short term):

- https://github.com/scipy/scipy/issues/24205

Since SciPy is used in many scikit-learn estimators and since this PR hasn't moved much since last year, I think we can also consider that dask support is not a short term priority for scikit-learn.

ogrisel commented:
Let's close for now. We can always reopen a new PR synchronized with `main` if array API support improves in dask and dask support improves in SciPy.
